utilities_common/multi_asic.py

```diff
@@ -0,0 +1,159 @@
+        For single asic, this function is not applicable
+        '''
+        if object_type == PORT_OBJ:
+            return is_port_internal(cli_object)
+        elif object_type == PORT_CHANNEL_OBJ:
+            return is_port_channel_internal(cli_object)
+        elif object_type == BGP_NEIGH_OBJ:
+            return is_bgp_session_internal(cli_object)
+
+    def skip_display(self, object_type, cli_object):
+        '''
+        The function determines if the passed cli_object has to be displayed or not.
+        returns true if the display_option is external and  the cli object is internal.
+        returns false, if the cli option is all or if it the platform is single ASIC.
+
+        '''
+        if not is_multi_npu():
+            return False
+        if self.display_option == DISPLAY_ALL:
+            return False
+        return self.is_object_internal(object_type, cli_object)
+
+    def get_ns_list_based_on_options(self):
+        ns_list = []
+        if not is_multi_npu():
+            return [DEFAULT_NAMESPACE]
+        else:
+            namespaces = get_all_namespaces()
+            if self.namespace_option is None:
+                if self.display_option ==  DISPLAY_ALL:
+                    ns_list = namespaces['front_ns'] + namespaces['back_ns']
+                else:
+                    ns_list = namespaces['front_ns']
+            else:
+                ns_list = [self.namespace_option]
+        return ns_list
+
+def multi_asic_ns_choices():
+    if not is_multi_npu() :
+        return []
+    choices =  get_namespaces()
+    return choices
+
+def multi_asic_display_choices():
+    if not is_multi_npu():
+        return [DISPLAY_ALL]
+    else:
+        return [DISPLAY_ALL, DISPLAY_EXTERNAL]
+
+def multi_asic_display_default_option():
+    if not is_multi_npu():
+        return DISPLAY_ALL
+    else:
+        return  DISPLAY_EXTERNAL
+
+
+
+_multi_asic_click_options = [
+    click.option('--display', '-d', 'display', default=multi_asic_display_default_option(), show_default=True, type=click.Choice(multi_asic_display_choices()), help='Show internal interfaces'),
+    click.option('--namespace', '-n', 'namespace', default=None, type=click.Choice(multi_asic_ns_choices()), show_default=True, help='Namespace name or all'),
+    #click.option('--namespace', '-n', 'namespace',  callback=multi_asic_ns_callback, show_default=True, help='Namespace name or all'),
+    ]
+
+def multi_asic_click_options(func):
+    for option in reversed(_multi_asic_click_options):
+        func = option(func)
+    return func
+
+def run_on_all_asics(func):
+    '''
+    This decorator is used on the CLI functions which needs to be
+    run on all the namespaces in the multi ASIC platform
+    The decorator loops through all the required namespaces,
+    for every iteration, it connects to all the DBs and provides an handle
+    to the wrapped function.
+
+    '''
+    @functools.wraps(func)
+    def wrapped_run_on_all_asics(self, *args, **kwargs):
```

Comment:
This is accessing class instance inside the decorator. Either decorator should be inside the CLI class, or we should come up with another way to run command in multiple namespaces.

Reply:
The decorator is for a function in the CLI class.

Reply:
Would prefer not changing the object state outside its domain. How about adding an API inside MultiAsic class, and call it from here.

```
def connected_dbs(ns):
            self.multi_asic.current_namespace = ns
            self.db = connect_to_all_dbs_for_ns(ns)
            self.config_db = connect_config_db_for_ns(ns)
```